### PR TITLE
Add options to disable trials deep copy.

### PR DESCRIPTION
**Pull request by sile**
This PR takes over #536.

This PR mainly contains the following changes:
- Add `deepcopy` option to `BaseStorage.get_all_trials()` method
- Add `Trial.get_trials(deepcopy: bool)` method
- Change `TPESampler` to use `get_trials(deepcopy=False)` when retrieving trials

### Performance comparison between this PR and `v0.18.1`

Benchmark code:
```python
# filename: optimize.py
import optuna
import datetime

optuna.logging.set_verbosity(optuna.logging.WARNING)

def objective(trial):
    x = trial.suggest_uniform('x', -10, 10)
    return x ** 2

study = optuna.create_study()

start = datetime.datetime.now()
study.optimize(objective, n_trials=1000, gc_after_trial=False)
print("Elapsed: {}".format(datetime.datetime.now() - start))
```

Run benchmark:
```console
// v0.18.1
$  pip uninstall -y optuna; pip install --user optuna==0.18.1
$ python /tmp/optimize.py
Elapsed: 0:00:20.021463  // 20 seconds

// This PR
$  pip uninstall -y optuna; pip install --user git+https://github.com/pfnet/optuna@followup-536
$ python /tmp/optimize.py
Elapsed: 0:00:01.991370  // 2 seconds
```

### CODE REVIEW

File: optuna/study.py [hunk 1]
```diff
@@ -105,11 +105,35 @@ def trials(self):
...
+
+        return self.get_trials()
+
+    def get_trials(self, deepcopy=True):
+        # type: (bool) -> List[structs.FrozenTrial]
+        """Return all trials in the study.
+
+        The returned trials are ordered by trial number.
+
+        For ordinal library users, it's recommended to use more handy
```
Comment (toshihikoyanase) [inline optuna/study.py]: I'm not sure, but it seems to be a typo. I think we can remove it and call them just `library users`.
```suggestion
        For ordinary library users, it's recommended to use more handy
```

Comment (sile) [reply optuna/study.py]: Thanks. Your suggestion makes sense. I'll remove the adjective.

Comment (sile) [reply optuna/study.py]: Fixed: https://github.com/optuna/optuna/pull/696/commits/84d4d838afdfa25054b88d6947a146143a4d5798

File: optuna/storages/base.py [hunk 2]
```diff
@@ -164,8 +165,8 @@ def get_trial(self, trial_id):
         raise NotImplementedError
 
     @abc.abstractmethod
-    def get_all_trials(self, study_id):
-        # type: (int) -> List[structs.FrozenTrial]
+    def get_all_trials(self, study_id, deepcopy=True):
```
Comment (crcrpar) [inline optuna/storages/base.py]: Why `deepcopy` defaults to `True`?

For backward-compatibility?

Comment (sile) [reply optuna/storages/base.py]: It's for safety.
Without deep-copying, the trials are shared by caller and callee. That is, if the caller (unintendedly) mutates a part of the returned trials, the consistency of the storage state could break (and vice versa).
When disabling deep-copying, the caller should be aware of the effect and risk, so I think the default value should be `False`.

File: optuna/storages/rdb/storage.py [hunk 3]
```diff
@@ -1059,15 +1067,18 @@ def is_empty(self):
...
+        # type: (structs.FrozenTrial, bool) -> None
 
         if not self._enabled:
             return
 
         if trial.state.is_finished():
             with self._lock:
-                self._finished_trials[trial._trial_id] = copy.deepcopy(trial)
+                if deepcopy:
+                    trial = copy.deepcopy(trial)
```
Comment (hvy) [inline optuna/storages/rdb/storage.py]: Considering how this method is called, i.e. with a trial that was just constructed from the DB, we can always create a shallow copy here (and we won't need the `deepcopy` argument for the method).

Comment (sile) [reply optuna/storages/rdb/storage.py]: You're right. Thanks!
I'll remove the argument.

Comment (sile) [reply optuna/storages/rdb/storage.py]: Fixed: https://github.com/optuna/optuna/pull/696/commits/339af5764727e08640747b03416c4a7741678965

Comment (hvy) [reply optuna/storages/rdb/storage.py]: Ah, we must change the two calls to this method excluding the argument too.

Comment (sile) [reply optuna/storages/rdb/storage.py]: Oops..., sorry. Fixed: https://github.com/optuna/optuna/pull/696/commits/0325b95874595ddbd942d930a35ad26e295fdbf1

Comment (hvy) [reply optuna/storages/rdb/storage.py]: Thanks for the swift fix!

**Pull request merged**